src/React/Hook.purs

```diff
@@ -0,0 +1,310 @@
+module React.Hook
+  ( Hook
+
+  , useState
+  , useStateLazy
+  , setState
+  , modifyState
+  , SetState
+
+  , useEffect
+  , effectInput
+  , EffectInput
+
+  , useContext
+  , Context
+
+  , useReducer
+  , useReducerLazy
+  , dispatch
+  , Dispatch
+
+  , useCallback
+  , callbackInput
+  , CallbackInput
+
+  , useMemo
+  , memoInput
+  , MemoInput
+
+  , useRef
+  , getRef
+  , setRef
+  , Ref
+
+  , useImperativeMethods
+  , imperativeMethodsInput
+  , ImperativeMethodsInput
+
+  , useMutationEffect
+  , mutationEffectInput
+  , MutationEffectInput
+
+  , useLayoutEffect
+  , layoutEffectInput
+  , LayoutEffectInput
+  ) where
+
+import Prelude
+
+import Data.Function.Uncurried (Fn1, Fn2, Fn3, Fn4, mkFn2, runFn1, runFn2, runFn3, runFn4)
+import Data.Maybe (Maybe)
+import Data.Nullable (Nullable)
+import Data.Nullable as Nullable
+import Data.Tuple (Tuple(..))
+
+import Effect (Effect)
+import Effect.Uncurried (EffectFn1, EffectFn2, runEffectFn1, runEffectFn2)
+
+import Unsafe.Coerce (unsafeCoerce)
+
+useState
+  :: forall a
+   . a
+  -> Hook (Tuple a (SetState a))
+useState = runFn2 useState_ Tuple
+
+useStateLazy
+  :: forall a
+   . (Unit -> a)
+  -> Hook (Tuple a (SetState a))
+useStateLazy = runFn2 useState_ Tuple
+
+setState
+  :: forall a
+   . SetState a
+  -> a
+  -> Effect Unit
+setState k = runEffectFn1 k'
+  where
+  k' :: EffectFn1 a Unit
+  k' = unsafeCoerce k
+
+modifyState
+  :: forall a
+   . SetState a
+  -> (a -> a)
+  -> Effect Unit
+modifyState k = runEffectFn1 k'
+  where
+  k' :: EffectFn1 (a -> a) Unit
+  k' = unsafeCoerce k
+
+foreign import data SetState :: Type -> Type
+
+foreign import useState_
+  :: forall a b
+   . Fn2 (b -> SetState b -> Tuple b (SetState b))
+         a
+         (Hook (Tuple b (SetState b)))
+
+useEffect
+  :: forall a
+   . Effect (Effect a)
+  -> Maybe (Array EffectInput)
```

Comment:
Is there a reason to not go ahead and lose the `Maybe`. There are two "empty" cases with `Just []`. If a user always has to pass in _something_ they might as well pass in `[]`.

Reply:
I believe that React behaves differently if you pass an empty array vs. not passing an array at all.

https://reactjs.org/docs/hooks-reference.html#conditionally-firing-an-effect
Not passing an array means that the effect is fired after every render. Passing an empty array means that the effect is fired only on mount and unmount.

Reply:
👍